charset/encoding_table.go

```diff
@@ -258,3 +259,26 @@ var encodings = map[string]struct {
 	"utf-16le":            {unicode.UTF16(unicode.LittleEndian, unicode.IgnoreBOM), "utf-16le"},
 	"x-user-defined":      {charmap.XUserDefined, "x-user-defined"},
 }
+
+// FindNextCharacterLength is used in lexer.peek() to determine the next character length.
+func FindNextCharacterLength(label string) func([]byte) int {
+	if f, ok := encodingNextCharacterLength[label]; ok {
+		return f
+	}
+	return nil
+}
+
+var encodingNextCharacterLength = map[string]func([]byte) int{
+	// https://en.wikipedia.org/wiki/GBK_(character_encoding)#Layout_diagram
+	"gbk": func(bs []byte) int {
+		if len(bs) == 0 || bs[0] < 0x80 {
+			// A byte in the range 00–7F is a single byte that means the same thing as it does in ASCII.
+			return 1
+		}
+		return 2
+	},
+	"utf-8": func(bs []byte) int {
+		_, size := utf8.DecodeRune(bs)
```

Comment:
`utf8.DecodeRune` is expensive, utf8 can use the first byte to determine the length.
```
b < 0x80  ==> 1
b < 0xe0  ==> 2
b < 0xf0  ==> 3
other        ==> 4
```